## **Problem:**

You are given the following dataset:
1. **Audible Data** : https://www.kaggle.com/datasets/snehangsude/audible-dataset

Your task is to:
- Find the problems with the datasets.
- Define the Data Quality Dimensions.
- Try to clean the datasets.

In [23]:
import pandas as pd
import numpy as np
import re

In [11]:
audible_df = pd.read_csv("dataset/audible_uncleaned.csv")
audible_df

,name,author,narrator,time,releasedate,language,stars,price
0,Geronimo Stilton #11 & #12,Writtenby:GeronimoStilton,Narratedby:BillLobely,2 hrs and 20 mins,04-08-08,English,5 out of 5 stars34 ratings,468.00
1,The Burning Maze,Writtenby:RickRiordan,Narratedby:RobbieDaymond,13 hrs and 8 mins,01-05-18,English,4.5 out of 5 stars41 ratings,820.00
2,The Deep End,Writtenby:JeffKinney,Narratedby:DanRussell,2 hrs and 3 mins,06-11-20,English,4.5 out of 5 stars38 ratings,410.00
3,Daughter of the Deep,Writtenby:RickRiordan,Narratedby:SoneelaNankani,11 hrs and 16 mins,05-10-21,English,4.5 out of 5 stars12 ratings,615.00
4,"The Lightning Thief: Percy Jackson, Book 1",Writtenby:RickRiordan,Narratedby:JesseBernstein,10 hrs,13-01-10,English,4.5 out of 5 stars181 ratings,820.00
...,...,...,...,...,...,...,...,...
87484,Last Days of the Bus Club,Writtenby:ChrisStewart,Narratedby:ChrisStewart,7 hrs and 34 mins,09-03-17,English,Not rated yet,596.00
87485,The Alps,Writtenby:StephenO'Shea,Narratedby:RobertFass,10 hrs and 7 mins,21-02-17,English,Not rated yet,820.00
87486,The Innocents Abroad,Writtenby:MarkTwain,Narratedby:FloGibson,19 hrs and 4 mins,30-12-16,English,Not rated yet,938.00
87487,A Sentimental Journey,Writtenby:LaurenceSterne,Narratedby:AntonLesser,4 hrs and 8 mins,23-02-11,English,Not rated yet,680.00


In [3]:
for col in audible_df[['name', 'author', 'narrator', 'language']]:
    print(f"{col}: {audible_df[col].nunique()}")

name: 82767
author: 48374
narrator: 29717
language: 36


In [4]:
audible_df.duplicated().sum()

np.int64(0)

In [5]:
audible_df['name'].duplicated().sum()

np.int64(4722)

In [6]:
audible_df[audible_df['name'].duplicated(keep=False)].sort_values(by='name')

,name,author,narrator,time,releasedate,language,stars,price
63965,"""Das Böse ist des Menschen beste Kraft""",Writtenby:ChristianLiederer,Narratedby:ThomasKrause,2 hrs and 24 mins,23-12-21,german,Not rated yet,233.00
63978,"""Das Böse ist des Menschen beste Kraft""",Writtenby:ChristianLiederer,Narratedby:ThomasKrause,2 hrs and 24 mins,14-12-21,german,Not rated yet,233.00
23134,#Girlboss,Writtenby:SophiaAmoruso,Narratedby:SarahJesAustell,4 hrs and 40 mins,06-08-15,English,4 out of 5 stars7 ratings,615.00
27243,#Girlboss,Writtenby:SophiaAmoruso,Narratedby:AudreyFiorini,4 hrs and 53 mins,26-09-18,french,Not rated yet,755.00
22306,100 quotes by Benjamin Franklin,Writtenby:BenjaminFranklin,Narratedby:PaulSpera,19 mins,07-03-18,English,5 out of 5 stars2 ratings,187.00
...,...,...,...,...,...,...,...,...
55220,途上,Writtenby:谷崎潤一郎,Narratedby:日下純,55 mins,01-02-22,japanese,Not rated yet,139.00
55548,雨ニモマケズ,Writtenby:宮沢賢治,Narratedby:孝介,1 min,24-06-21,japanese,Not rated yet,223.00
54181,雨ニモマケズ,Writtenby:宮沢賢治,Narratedby:宮沢りえ,1 min,16-11-15,japanese,Not rated yet,279.00
54172,雨ニモマケズ,Writtenby:宮澤賢治,Narratedby:西靖,1 min,15-06-17,japanese,Not rated yet,139.00


In [7]:
audible_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 87489 entries, 0 to 87488
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   name         87489 non-null  object
 1   author       87489 non-null  object
 2   narrator     87489 non-null  object
 3   time         87489 non-null  object
 4   releasedate  87489 non-null  object
 5   language     87489 non-null  object
 6   stars        87489 non-null  object
 7   price        87489 non-null  object
dtypes: object(8)
memory usage: 5.3+ MB


In [8]:
audible_df.describe()

,name,author,narrator,time,releasedate,language,stars,price
count,87489,87489,87489,87489,87489,87489,87489,87489
unique,82767,48374,29717,2284,5058,36,665,1011
top,The Art of War,"Writtenby:矢島雅弘,石橋遊",Narratedby:anonymous,2 mins,16-05-18,English,Not rated yet,586.00
freq,20,874,1034,372,773,61884,72417,5533


In [48]:
audible_df.sample(8)

,name,author,narrator,time,releasedate,language,stars,price
27668,The Leadership Secrets of Colin Powell,Writtenby:OrenHarari,Narratedby:ChrisRyan,4 hrs and 24 mins,21-07-21,English,Not rated yet,469.00
81228,The National Forgotten League,Writtenby:DanDaly,Narratedby:MichaelButlerMurray,14 hrs and 4 mins,15-07-14,English,Not rated yet,836.00
54499,Riviera - Der Weg in die Freiheit,Writtenby:JuliaKröhn,Narratedby:CharlottePuder,12 hrs and 30 mins,27-11-21,german,Not rated yet,669.00
44104,Bound Forever,Writtenby:AvaMarch,Narratedby:DavidThorpe,3 hrs and 31 mins,30-01-17,English,Not rated yet,501.00
9884,Robo-Sauce,Writtenby:AdamRubin,Narratedby:AdamRubin,6 mins,26-10-21,English,Not rated yet,251.00
5803,All Because You Matter,Writtenby:TamiCharles,Narratedby:TamiCharles,16 mins,06-10-20,English,Not rated yet,164.00
47365,Collectibles,Writtenby:LawrenceBlock-editor,"Narratedby:PeterBerkrot,TeriSchnaubelt",10 hrs and 29 mins,01-02-22,English,Not rated yet,703.00
14578,Glitch Feminism,Writtenby:LegacyRussell,Narratedby:JaninaEdwards,3 hrs and 4 mins,13-11-20,English,5 out of 5 stars1 rating,351.00


## Dataset Summary
This dataset has been collected using **Web Scraping from Audible** and contains all the basic information of the **87489 Audiobooks**.
- Total Audiobooks: 87489
- Unique Audiobooks: 82767
    - > Apparently there are 87489-82767=**4722** duplicated names of audiobooks, all having some differences. For example, some have different *narrator*, some have different *language*, some have different *time*, some have different *price*, etc.
- Unique Authors: 48374
- Unique Narrator: 29717
- Unique Language: 36

### Column Descriptions
- `name`: Names of the Audiobook
- `author`: Names of author of the original book
- `narrator`: Names of the narrator of the Audiobook
- `time`: Duration of the Audiobook's time
- `releasedate`: Date of release of the Audibook
- `language`: Language of the Audiobook
- `stars`: Average Star rating of the Audiobook by the listeners out of 5, and rating count
- `price`: Price of the Audiobook

## Issues Identified
### Quality issues
- `name` contains symbols like '#', ' " ', etc
- `author` has prefix "**Writtenby:**" attached to every value
- `author` has name written in Pascal Case
- `author` has names of authors seperated by commas
- `narrator` has prefix "**Narratedby:**" attached to every value
- `narrator` has name written in Pascal Case
- `narrator` has names of narrators seperated by commas
- `time` is not a `datetimeobject`
- `releasedate` is not a `datetimeobject`
- `language` must be a categorical type
- `stars` contains quantitative data stored as a non-numeric type
- `stars` also contains the review count
- `price` is an object instead of float

### **Consistency** (formatting, representation, standardization issues)

These problems arise when the same type of data is not represented uniformly:

* `name` contains symbols like `#`, `"`, etc
* `author` has prefix "**Writtenby:**" attached
* `author` has name written in Pascal Case
* `author` has names separated by commas
* `narrator` has prefix "**Narratedby:**" attached
* `narrator` has name written in Pascal Case
* `narrator` has names separated by commas

These are all about inconsistent formatting or representation rather than correctness.

---

### **Validity** (data type / rule violations)

These issues break predefined rules, schemas, or expected data types:

* `time` is not a datetime object
* `releasedate` is not a datetime object
* `language` must be a categorical type
* `stars` contains quantitative data stored as a non-numeric type
* `price` is an object instead of float

These violate structural or schema expectations.

---

### **Accuracy** (incorrect or misleading representation of real-world meaning)

These issues involve values that don’t correctly reflect what they’re supposed to represent:

* `stars` also contains the review count

This mixes two different real-world concepts (rating vs. count), making the data misleading.

---

### **Completeness**

* No issues from the list fall under completeness (nothing explicitly missing).

---

### Quick Summary

| Dimension    | Issues Count | Core Problem Type             |
| ------------ | ------------ | ----------------------------- |
| Consistency  | 7            | Formatting / representation   |
| Validity     | 5            | Data type / schema violations |
| Accuracy     | 1            | Misrepresentation of meaning  |
| Completeness | 0            | Missing data                  |

Then the order of severity is:
`Completeness > Validity > Accuracy > Consistency`

In [9]:
audible_df.head()

,name,author,narrator,time,releasedate,language,stars,price
0,Geronimo Stilton #11 & #12,Writtenby:GeronimoStilton,Narratedby:BillLobely,2 hrs and 20 mins,04-08-08,English,5 out of 5 stars34 ratings,468.00
1,The Burning Maze,Writtenby:RickRiordan,Narratedby:RobbieDaymond,13 hrs and 8 mins,01-05-18,English,4.5 out of 5 stars41 ratings,820.00
2,The Deep End,Writtenby:JeffKinney,Narratedby:DanRussell,2 hrs and 3 mins,06-11-20,English,4.5 out of 5 stars38 ratings,410.00
3,Daughter of the Deep,Writtenby:RickRiordan,Narratedby:SoneelaNankani,11 hrs and 16 mins,05-10-21,English,4.5 out of 5 stars12 ratings,615.00
4,"The Lightning Thief: Percy Jackson, Book 1",Writtenby:RickRiordan,Narratedby:JesseBernstein,10 hrs,13-01-10,English,4.5 out of 5 stars181 ratings,820.00


In [15]:
# remove whitespaces prefix and whitespaces from col 'author' values
audible_df['author'] = audible_df['author'].str.replace('Writtenby:', '', regex=False).str.replace(' ', '', regex=False)

# ensure it works
audible_df['author'].head().to_list()

['GeronimoStilton', 'RickRiordan', 'JeffKinney', 'RickRiordan', 'RickRiordan']

In [18]:
# turn pasal case of col 'author' into title case
# Insert a space before capital letters (ignoring the start of the string)
audible_df['author'] = audible_df['author'].str.replace(r'(?<!^)(?=[A-Z])', ' ', regex=True)

# ensure it works
audible_df['author'].head().to_list()

['Geronimo Stilton',
 'Rick Riordan',
 'Jeff Kinney',
 'Rick Riordan',
 'Rick Riordan']

In [16]:
# remove whitespaces prefix and whitespaces from col 'narrator' values
audible_df['narrator'] = audible_df['narrator'].str.replace('Narratedby:', '', regex=False).str.replace(' ', '', regex=False)

# ensure it works
audible_df['narrator'].head().to_list()

['BillLobely',
 'RobbieDaymond',
 'DanRussell',
 'SoneelaNankani',
 'JesseBernstein']

In [19]:
# turn pasal case of col 'author' into title case
# Insert a space before capital letters (ignoring the start of the string)
audible_df['narrator'] = audible_df['narrator'].str.replace(r'(?<!^)(?=[A-Z])', ' ', regex=True)

# ensure it works
audible_df['narrator'].head().to_list()

['Bill Lobely',
 'Robbie Daymond',
 'Dan Russell',
 'Soneela Nankani',
 'Jesse Bernstein']

In [ ]:
# convert 'time' to Timedelta object
def parse_time(x):
    if pd.isna(x):
        return pd.NaT
    
    # Handle special case
    if 'less than' in x.lower():
        return pd.Timedelta(minutes=1)
    
    hrs = 0
    mins = 0
    
    hr_match = re.search(r'(\d+)\s*hr', x)
    min_match = re.search(r'(\d+)\s*min', x)
    
    if hr_match:
        hrs = int(hr_match.group(1))
    if min_match:
        mins = int(min_match.group(1))
    
    return pd.Timedelta(hours=hrs, minutes=mins)

audible_df['time'] = audible_df['time'].apply(parse_time)
audible_df['time']

0       0 days 02:20:00
1       0 days 13:08:00
2       0 days 02:03:00
3       0 days 11:16:00
4       0 days 10:00:00
              ...      
87484   0 days 07:34:00
87485   0 days 10:07:00
87486   0 days 19:04:00
87487   0 days 04:08:00
87488   0 days 06:01:00
Name: time, Length: 87489, dtype: timedelta64[ns]

In [ ]:
# convert 'releasedate' to datetime object
audible_df['releasedate'] = pd.to_datetime(audible_df['releasedate'], format='%d-%m-%y')
audible_df['releasedate']

0       2008-08-04
1       2018-05-01
2       2020-11-06
3       2021-10-05
4       2010-01-13
           ...    
87484   2017-03-09
87485   2017-02-21
87486   2016-12-30
87487   2011-02-23
87488   2017-03-07
Name: releasedate, Length: 87489, dtype: datetime64[ns]

In [66]:
# convert 'language' to 'category' type
audible_df['language'] = audible_df['language'].astype('category')

In [74]:
# Extract rating (e.g., 5, 4.5)
audible_df['star_rating'] = audible_df['stars'].str.extract(r'(\d+\.?\d*)').astype(float)

# Extract number of ratings (e.g., 34)
audible_df['rating_count'] = audible_df['stars'].str.extract(r'(\d+)\s*ratings').fillna(0).astype('int')

# Drop original column stars, no longer needed
audible_df = audible_df.drop(columns=['stars'])

In [84]:
# convert 'price' to float
audible_df['price'] = pd.to_numeric(audible_df['price'].str.replace(',', ''), errors='coerce')
audible_df['price']

0        468.0
1        820.0
2        410.0
3        615.0
4        820.0
         ...  
87484    596.0
87485    820.0
87486    938.0
87487    680.0
87488    569.0
Name: price, Length: 87489, dtype: float64

In [85]:
audible_df

,name,author,narrator,time,releasedate,language,price,star_rating,rating_count
0,Geronimo Stilton #11 & #12,Geronimo Stilton,Bill Lobely,0 days 02:20:00,04-08-08,English,468.0,5.0,34
1,The Burning Maze,Rick Riordan,Robbie Daymond,0 days 13:08:00,01-05-18,English,820.0,4.5,41
2,The Deep End,Jeff Kinney,Dan Russell,0 days 02:03:00,06-11-20,English,410.0,4.5,38
3,Daughter of the Deep,Rick Riordan,Soneela Nankani,0 days 11:16:00,05-10-21,English,615.0,4.5,12
4,"The Lightning Thief: Percy Jackson, Book 1",Rick Riordan,Jesse Bernstein,0 days 10:00:00,13-01-10,English,820.0,4.5,181
...,...,...,...,...,...,...,...,...,...
87484,Last Days of the Bus Club,Chris Stewart,Chris Stewart,0 days 07:34:00,09-03-17,English,596.0,NaN,0
87485,The Alps,Stephen O' Shea,Robert Fass,0 days 10:07:00,21-02-17,English,820.0,NaN,0
87486,The Innocents Abroad,Mark Twain,Flo Gibson,0 days 19:04:00,30-12-16,English,938.0,NaN,0
87487,A Sentimental Journey,Laurence Sterne,Anton Lesser,0 days 04:08:00,23-02-11,English,680.0,NaN,0


In [92]:
audible_df.to_csv("dataset/audible_cleaned.csv")